controllers/group_totals.go

```diff
@@ -56,48 +54,132 @@ func InstaCountHandler(rw http.ResponseWriter, r *http.Request) {
-	decoder := json.NewDecoder(r.Body)
-	group := &fhirmodels.Group{}
-	err := decoder.Decode(group)
-	if err != nil {
-		http.Error(rw, err.Error(), http.StatusInternalServerError)
+	pquery = TrimSuffix(pquery, "&")
+	cquery = TrimSuffix(cquery, "&")
+	equery = TrimSuffix(equery, "&")
+
+	searcher := search.NewMongoSearcher(server.Database)
+
+	var pResultIDs []struct {
+		ID string `bson:"_id"`
+	}
+	pSearchQuery := search.Query{Resource: "Patient", Query: pquery}
+	pQ := searcher.CreateQuery(pSearchQuery)
+	pQ.Select(bson.M{"_id": 1}).All(&pResultIDs)
+	pids := make([]string, len(pResultIDs))
+	for i := range pResultIDs {
+		pids[i] = pResultIDs[i].ID
 	}
 
-	pipelineMap := make(map[string]models.Pipeline)
+	if intersectRequired {
+		type patientContainer struct {
+			ID string `bson:"referenceid"`
+		}
 
-	pipelineMap["patients"] = models.NewPipeline(group)
-	pipelineMap["encounters"] = models.NewEncounterPipeline(group)
-	pipelineMap["conditions"] = models.NewConditionPipeline(group)
+		var cResultIDs []struct {
+			Patient patientContainer `bson:"patient"`
+		}
+		cSearchQuery := search.Query{Resource: "Condition", Query: cquery}
+		cQ := searcher.CreateQuery(cSearchQuery)
+		cQ.Select(bson.M{"patient.referenceid": 1}).All(&cResultIDs)
+		cids := make([]string, len(cResultIDs))
+		for i := range cResultIDs {
+			cids[i] = cResultIDs[i].Patient.ID
+		}
 
-	resultMap := make(map[string]int)
+		var eResultIDs []struct {
+			Patient patientContainer `bson:"patient"`
+		}
+		eSearchQuery := search.Query{Resource: "Encounter", Query: equery}
+		eQ := searcher.CreateQuery(eSearchQuery)
+		eQ.Select(bson.M{"patient.referenceid": 1}).All(&eResultIDs)
+		eids := make([]string, len(eResultIDs))
+		for i := range eResultIDs {
+			eids[i] = eResultIDs[i].Patient.ID
+		}
 
-	for pipelineType, pipeline := range pipelineMap {
-		qr, err := pipeline.ExecuteCount(server.Database)
-		if err != nil && err != mgo.ErrNotFound {
-			http.Error(rw, err.Error(), http.StatusInternalServerError)
-			return
+		firstIntersect := make(map[string]bool)
+		secondIntersect := make(map[string]bool)
+		finalIntersect := make(map[string]bool)
+
+		for _, pid := range pids {
+			firstIntersect[pid] = true
+		}
+
+		for _, cid := range cids {
+			if firstIntersect[cid] == true {
+				secondIntersect[cid] = true
+			}
+		}
+
+		for _, eid := range eids {
+			if secondIntersect[eid] == true {
+				finalIntersect[eid] = true
+			}
+		}
+
+		intersectPIDs := make([]string, 0, len(finalIntersect))
+		for id := range finalIntersect {
+			intersectPIDs = append(intersectPIDs, id)
 		}
-		resultMap[pipelineType] = qr.Total
+		pids = intersectPIDs
+	}
+
+	pCount := len(pids)
+
+	cCollection := server.Database.C("conditions")
+	cCount, err := cCollection.Find(bson.M{"patient.referenceid": bson.M{"$in": pids}}).Count()
+
+	eCollection := server.Database.C("encounters")
+	eCount, err := eCollection.Find(bson.M{"patient.referenceid": bson.M{"$in": pids}}).Count()
+
```

Comment:
Not a super big deal, but it strikes me that we're mixing the FHIR search library and direct Mongo access together here.  We may want to think about if we want to use _only_ the FHIR search library, or if we want to use _only_ MGO directly.  There are probably advantages / disadvantages to each.